### DISCUSSION TITLE
Javascript Rules via OpenJDK Nashorn

### DISCUSSION THREAD

**Discussion by justinstroud-campmeeker** [pmd/pmd#5193] [Ideas]
Hi all. In the regular expression discussion, I got a bit sidetracked (big surprise) but I think that sidetrack is worth following if only for discussion purposes.

After some investigation and a bit of PoC testing, I was able to build a generic "JavascriptRule" that passes the visited node to a script and gets information back as to whether or not to flag a violation. There are no wrappers or custom abstractions written around the AST classes. They just work out of the box. Same thing with node stream methods (`descendants()`, for example.)

The only dependency is on the OpenJDK implementation of the Nashorn Javascript engine for Java:
```xml
        <dependency>
            <groupId>org.openjdk.nashorn</groupId>
            <artifactId>nashorn-core</artifactId>
            <version>15.4</version>
        </dependency>
```

Here is a gist with some of the code ripped from my PoC:
https://gist.github.com/justinstroudbah/17790f569c731d988ec62b238b93d7df/edit

I have some other examples as well. I rewrote some custom rules that I've deployed over the past few months, and although there are some quirky differences between Java and Javascript (with the latter being dynamic) it's really not that difficult or even hacky.

I'd love to get a discussion going around this, because I think it would be a great feature (if only for my own custom rule library.) For example, most Salesforce folks aren't that well versed in XPath and it's pretty rare to have a company that lets them split time between Apex and Java. But Javascript? That's their client side. That's LWC development. It's right in their wheelhouse.

**Comment by justinstroud-campmeeker** [comment]
I had to give myself a refresher on Nashorn (I've worked with it before in a completely different context...software DSP.) This article was mighty helpful:
https://winterbe.com/posts/2014/04/05/java8-nashorn-tutorial/

**Comment by justinstroud-campmeeker** [comment]
Honestly, I'm pretty close to just cleaning my work up and creating a PR out of it. Let the chips fall where they may. The only hitch is the Maven dependency, but that will be part of the commit.

**Comment by oowekyala** [comment]
I think that's nice but should be outside of core because of the dependency. Maybe in a new module pmd-javascript-extensions (name is open to bikeshedding)

**Reply by justinstroud-campmeeker** [reply]
Thanks @oowekyala  I was thinking the same thing. I took a look at the XPath rule as a starting point, but a sep. module is probably the best idea? Anything with core seems pretty overwhelming right now.

The other option is to start with something I'm more familiar with, the Apex module. Build out the JavaScript rule there.

The only issue is this: if I build a rule but that rule has multiple different entries in my rule XML (with different names and different properties, just pointing to the same Java class) will that work? I seem to remember something about PMD instantiating a rule only once per file but I don't know if that based on the class or the  name of the rule because those are two different things.

EDIT: I have two approaches in mind as well. Abstract class in its own module that has all the important stuff for scripting, then each module can inherit from it and take advantage in their own way. This would be pretty dependent on consistent language implementations though :) The other approach, the one with the smallest footprint and the more prudent route is to do the above, build out a "JavaScipted" rule in the Apex module for the community to use.

**Reply by oowekyala** [reply]
> The other option is to start with something I'm more familiar with, the Apex module. Build out the JavaScript rule there.

I think that has the same problems as putting it into core. That would add a dependency to the apex module that many won't need to use.

I would suggest putting your rule in its own module. It doesn't seem like it is language-specific and I don't think we need language-specific base classes. 

If you want to build some Apex-specific things, we can make some room in the apex module and add a provided dependency on the js-scripting module. (maybe `pmd-js-scripting` is a good name for the module btw). But I can't really see what things would be both Apex-specific and specific to your Javascript rules... Would that be javascript code?

**Reply by justinstroud-campmeeker** [reply]
Hi @oowekyala thank you for the response. That totally makes sense. I did throw together a better proof-of-concept, and called the (generically) JavaScripted rules as opposed to JavaScript rules so as not to confuse with the existing language module. I will add you as a contributor to the repo (it's private currently) so you can check it out. Be on the lookout for an invite! Again, this is a PoC and I want to keep it as private until there is a proper home for it. I'm looking at ways that the idea can be leveraged in the same way custom rules are for the Salesforce community, which means configuring Maven to just include the Nashorn dependency in the custom rule jar.

At any rate, take a look when you have a chance and let me know what you think.
